File: samples/catalog-ios-app/iosApp/Loggers/CrashlyticsLogger.swift
//
//  CrashlyticsLogger.swift
//  iosApp
//
//  Created by Claude AI on 13/08/2025.
//  Copyright © 2025 orgName. All rights reserved.
//

import Foundation
import FirebaseCrashlytics
import appShared

class CrashlyticsLogger: ICrashLogger {
    private var crashlytics: Crashlytics {
        return Crashlytics.crashlytics()
    }
    
    func setCurrentScreen(screenName: String) {
        crashlytics.setCustomValue(screenName, forKey: "CurrentScreen")
    }
    
    func logEvent(eventName: String, parameters: [String : Any]) {
        if !isEnhancedLoggingEnabled() { return }
        
        if parameters.isEmpty {
            crashlytics.log(eventName)
        } else {
            crashlytics.log("\(eventName): \(parameters)")
        }
    }
    
    func logException(exception: KotlinThrowable) {
        if !isEnhancedLoggingEnabled() { return }
        
        let error = NSError(domain: "KotlinException", code: 0, userInfo: [
            NSLocalizedDescriptionKey: exception.message ?? "Unknown Kotlin exception"
        ])
        crashlytics.record(error: error)
    }
    
    func setUserId(userId: String) {
        crashlytics.setUserID(userId)
    }
    
    func setCustomKey(key: String, value: String) {
        crashlytics.setCustomValue(value, forKey: key)
    }
    
    func clear() {
        crashlytics.setUserID("")
    }
    
    private func isEnhancedLoggingEnabled() -> Bool {
        return DiProvider.remoteConfigs.getBoolean(key: AppConstants().CONFIG_ENABLE_ENHANCED_CRASHLYTICS)?.boolValue ?? false
    }
}

File: samples/catalog-app-shared/src/androidMain/kotlin/com/metacto/catalogapp/loggers/CrashlyticsLogger.kt
package com.metacto.catalogapp.loggers

import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.metacto.catalogapp.constants.AppConstants
import com.metacto.core.extensions.orFalse
import com.metacto.kmm.remoteconfig.common.RemoteConfigProvider
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

class CrashlyticsLogger : ICrashLogger, KoinComponent {
    private val remoteConfigs by inject<RemoteConfigProvider>()
    private val crashlytics by lazy {
        FirebaseCrashlytics.getInstance()
    }

    override fun setCurrentScreen(screenName: String) {
        crashlytics.setCustomKey("CurrentScreen", screenName)
    }

    override fun logEvent(eventName: String, parameters: Map<String, Any>) {
        if (!isEnhancedLoggingEnabled()) return

        if (parameters.isEmpty()) {
            crashlytics.log(eventName)
        } else {
            crashlytics.log("$eventName: $parameters")
        }
    }

    override fun logException(exception: Throwable) {
        if (!isEnhancedLoggingEnabled()) return
        crashlytics.recordException(exception)
    }

    override fun setUserId(userId: String) {
        crashlytics.setUserId(userId)
    }

    override fun setCustomKey(key: String, value: String) {
        crashlytics.setCustomKey(key, value)
    }

    override fun clear() {
        crashlytics.setUserId("")
    }

    private fun isEnhancedLoggingEnabled(): Boolean {
        return remoteConfigs.getBoolean(AppConstants.CONFIG_ENABLE_ENHANCED_CRASHLYTICS).orFalse()
    }
}